kolibri/plugins/coach/api.py

```diff
@@ -188,57 +189,117 @@ def list(self, request, *args, **kwargs):
         except (FacilityUser.DoesNotExist, Collection.DoesNotExist, ValueError):
             return False
 
 
-@query_params_required(collection_id=str)
-class BaseDifficultQuestionsViewset(viewsets.ViewSet):
-    permission_classes = (permissions.IsAuthenticated, DifficultiesPermissions,)
+# Define a base class so that the inherited class is properly introspectable,
+# rather than being the result of our wrapping
+@query_params_required(classroom_id=str)
+class BaseExerciseDifficultQuestionsViewset(viewsets.ViewSet):
+    pass
 
 
-class ExerciseDifficultQuestionsViewset(BaseDifficultQuestionsViewset):
+class ExerciseDifficultQuestionsViewset(BaseExerciseDifficultQuestionsViewset):
+    permission_classes = (permissions.IsAuthenticated, ExerciseDifficultiesPermissions,)
 
     def retrieve(self, request, pk):
         """
         Get the difficult questions for a particular exercise.
         pk maps to the content_id of the exercise in question.
         """
-        collection_id = self.kwargs['collection_id']
+        classroom_id = request.GET.get('classroom_id', None)
+        group_id = request.GET.get('group_id', None)
+        lesson_id = request.GET.get('lesson_id', None)
         queryset = AttemptLog.objects.filter(masterylog__summarylog__content_id=pk)
-        queryset = HierarchyRelationsFilter(queryset).filter_by_hierarchy(
-            ancestor_collection=collection_id,
-            target_user=F("user"),
-        )
+        if lesson_id is not None:
+            collection_ids = Lesson.objects.get(id=lesson_id).lesson_assignments.values_list('collection_id', flat=True)
+            if group_id is not None:
+                if group_id not in collection_ids and classroom_id not in collection_ids:
+                    # In the special case that the group is not in the lesson assignments
+                    # nor the containing classroom, just return an empty queryset.
+                    queryset = AttemptLog.objects.none()
+            else:
+                # Only filter by all the collections in the lesson if we are not also
+                # filtering by a specific group. Otherwise the group should be sufficient.
+                base_queryset = queryset
+                # Set starting queryset to null, then OR.
+                queryset = AttemptLog.objects.none()
+                for collection_id in collection_ids:
+                    queryset |= HierarchyRelationsFilter(base_queryset).filter_by_hierarchy(
+                        ancestor_collection=collection_id,
+                        target_user=F("user"),
+                    )
+                queryset = queryset.distinct()
+        if group_id is not None:
+            collection_id = group_id or classroom_id
+            queryset = HierarchyRelationsFilter(queryset).filter_by_hierarchy(
+                ancestor_collection=collection_id,
+                target_user=F("user"),
+            )
+
```

Comment:
Lots going on here that I don't understand! but it looks tested so I trust it

Reply:
It's the only reason that I do.